module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/AlmaTest.php

```diff
@@ -0,0 +1,243 @@
+
+    /**
+     * Standard setup method.
+     *
+     * @return void
+     */
+    public function setUp(): void
+    {
+        $this->driver = new Alma(new \VuFind\Date\Converter());
+    }
+
+    /**
+     * Replace makeRequest to inject test returns
+     *
+     * @param string        $path          Path to retrieve from API (excluding base
+     *                                     URL/API key)
+     * @param array         $paramsGet     Additional GET params
+     * @param array         $paramsPost    Additional POST params
+     * @param string        $method        GET or POST. Default is GET.
+     * @param string        $rawBody       Request body.
+     * @param Headers|array $headers       Add headers to the call.
+     * @param array         $allowedErrors HTTP status codes that are not treated as
+     *                                     API errors.
+     * @param bool          $returnStatus  Whether to return HTTP status in addition
+     *                                     to the response.
+     *
+     * @return null|\SimpleXMLElement
+     */
+    public function mockMakeRequest(
+        $path,
+        $paramsGet = [],
+        $paramsPost = [],
+        $method = 'GET',
+        $rawBody = null,
+        $headers = null,
+        $allowedErrors = [],
+        $returnStatus = false
+    ) {
+        // Get the next step of the test, and make assertions as necessary
+        // (we'll skip making assertions if the next step is empty):
+        $testData = $this->fixtureSteps[$this->currentFixtureStep] ?? [];
+        $this->currentFixtureStep++;
+        unset($testData['comment']);
+        if (!empty($testData['expectedParams'])) {
+            $msg = "Error in step {$this->currentFixtureStep} of fixture: "
+                . $this->currentFixture;
+            $params = func_get_args();
+            foreach ($testData['expectedParams'] as $i => $expectedParam) {
+                $this->assertEquals(
+                    $expectedParam,
+                    $params[$i],
+                    $msg . ", parameter index $i"
+                );
+            }
+        }
+
+        $resultFixture = $this->getFixture('alma/responses/' . $testData['resultFixture']);
+        $result = simplexml_load_string($resultFixture);
+        return $returnStatus
+            ? [$result, $testData['status'] ?? 200]
+            : $result;
+    }
+
+    /**
+     * Generate a new driver to return responses set in a json fixture
+     *
+     * Overwrites $this->driver
+     *
+     * @param string $test   Name of test fixture to load
+     * @param array  $config Driver configuration (null to use default)
+     *
+     * @return void
+     */
+    protected function createConnector(string $test, array $config = null): void
+    {
+        // Setup test responses
+        $this->fixtureSteps = $this->getJsonFixture("alma/responses/$test.json");
+        $this->currentFixture = $test;
+        $this->currentFixtureStep = 0;
```

Comment:
Now that we're using this pattern in three drivers, I wonder if we should consider making it a trait somehow. Not something we have to do right now, just food for thought...

Reply:
I tried to think of something clear and useful, but since the actual implementations vary a bit, I didn't come up with anything.